File: web/counter.js
/**
 * counter.js — Contador bidireccional de cruce de línea
 *
 * Port directo de la lógica Python (counter.py) a JavaScript.
 * Detecta cuándo un objeto cruza un segmento orientable:
 *   - Lado negativo → positivo = "in"  (Entrada)
 *   - Lado positivo → negativo = "out" (Salida)
 */

'use strict';

/** Categorías rastreadas (mismas que en config.py) */
const CATEGORY_KEYS = ['persons', 'bicycles', 'cars', 'motorcycles', 'buses', 'trucks'];


class Counter {
  /**
   * @param {Object}   opts
   * @param {number}   opts.linePosition  Posición relativa X del centro de línea (0.0 – 1.0)
   * @param {number}   opts.lineYPosition Posición relativa Y del centro de segmento (0.0 – 1.0)
   * @param {number}   opts.lineHeight    Longitud relativa del segmento (0.0 – 1.0)
   * @param {number}   opts.lineAngle     Ángulo en grados (0 = vertical)
   * @param {number}   opts.margin        Píxeles de margen para resetear la bandera de cruce
   * @param {Function} opts.onCross       Callback(categoria, direction, trackId)
   */
  constructor({ linePosition = 0.5, lineYPosition = 0.5, lineHeight = 1.0, lineAngle = 0, margin = 30, onCross = null } = {}) {
    this.linePos     = linePosition;
    this.lineYPos    = lineYPosition;
    this.lineHeightRel = lineHeight;
    this.lineAngleDeg = lineAngle;
    this.margin      = margin;
    this.onCross     = onCross;
    this.frameWidth  = 640;
    this.frameHeight = 480;
    this.lineX       = Math.round(this.frameWidth * this.linePos);
    this.lineY       = Math.round(this.frameHeight * this.lineYPos);
    this.lineLengthPx = Math.round(this.frameHeight * this.lineHeightRel);
    this.lineDirX = 0;
    this.lineDirY = 1;
    this.lineNormalX = 1;
    this.lineNormalY = 0;
    this.lineX1 = this.lineX;
    this.lineY1 = 0;
    this.lineX2 = this.lineX;
    this.lineY2 = this.frameHeight;
    this.lineTop     = 0;
    this.lineBottom  = this.frameHeight;
    this._recomputeLineSegment();

    /** Estado de cruce por track: Map<trackId, { categoria, canCross }> */
    this._trackState = new Map();

    /** Contadores: { persons: { in: 0, out: 0 }, ... } */
    this.counts = {};
    CATEGORY_KEYS.forEach(k => { this.counts[k] = { in: 0, out: 0 }; });
  }

  /* ── Configuración ──────────────────────────────────────────────────── */

  /** Actualiza dimensiones del frame (llamar al abrir cámara) */
  setFrameSize(w, h) {
    this.frameWidth  = w;
    this.frameHeight = h;
    this._recomputeLineSegment();
  }

  /** Actualiza posición relativa de la línea (0.02 – 0.98) */
  setLinePosition(pos) {
    this.linePos = Math.max(0.02, Math.min(0.98, pos));
    this._recomputeLineSegment();
  }

  /** Actualiza posición vertical relativa del centro del segmento (0.02 – 0.98) */
  setLineYPosition(pos) {
    this.lineYPos = Math.max(0.02, Math.min(0.98, pos));
    this._recomputeLineSegment();
  }

  /** Actualiza altura relativa del segmento (0.05 – 1.0) */
  setLineHeight(heightRel) {
    this.lineHeightRel = Math.max(0.05, Math.min(1.0, heightRel));
    this._recomputeLineSegment();
  }

  /** Actualiza ángulo del segmento en grados (-89 – 89, 0 = vertical) */
  setLineAngle(angleDeg) {
    this.lineAngleDeg = Math.max(-89, Math.min(89, angleDeg));
    this._recomputeLineSegment();
  }

  _recomputeLineSegment() {
    this.lineX = Math.round(this.frameWidth * this.linePos);
    this.lineY = Math.round(this.frameHeight * this.lineYPos);
    this.lineLengthPx = Math.round(this.frameHeight * this.lineHeightRel);

    const rad = (this.lineAngleDeg * Math.PI) / 180;
    this.lineDirX = Math.sin(rad);
    this.lineDirY = Math.cos(rad);

    // Normal para determinar lado de cruce. Con ángulo 0 queda: dist = x - lineX.
    this.lineNormalX = this.lineDirY;
    this.lineNormalY = -this.lineDirX;

    const half = this.lineLengthPx / 2;
    this.lineX1 = this.lineX - this.lineDirX * half;
    this.lineY1 = this.lineY - this.lineDirY * half;
    this.lineX2 = this.lineX + this.lineDirX * half;
    this.lineY2 = this.lineY + this.lineDirY * half;

    // Compatibilidad para cualquier uso legado del tramo vertical
    this.lineTop = Math.max(0, Math.min(this.frameHeight, Math.round(Math.min(this.lineY1, this.lineY2))));
    this.lineBottom = Math.max(0, Math.min(this.frameHeight, Math.round(Math.max(this.lineY1, this.lineY2))));
  }

  _signedDistanceToLine(point) {
    const dx = point[0] - this.lineX;
    const dy = point[1] - this.lineY;
    return dx * this.lineNormalX + dy * this.lineNormalY;
  }

  _lineProjection(point) {
    const dx = point[0] - this.lineX;
    const dy = point[1] - this.lineY;
    return dx * this.lineDirX + dy * this.lineDirY;
  }

  /* ── Procesamiento ──────────────────────────────────────────────────── */

  /**
   * Procesa detecciones con tracking.
    * Cada item debe tener: trackId, categoria, history (array de centros [x, y])
   * @param {Array<Object>} trackedDetections
   */
  process(trackedDetections) {
    const activeIds = new Set();

    for (const det of trackedDetections) {
      const { trackId, categoria, history } = det;
      if (!trackId || !this.counts[categoria]) continue;

      activeIds.add(trackId);

      // Inicializar estado si es nuevo
      if (!this._trackState.has(trackId)) {
        this._trackState.set(trackId, { categoria, canCross: true });
      }

      const state = this._trackState.get(trackId);

      // Comprobar cruce con al menos 2 puntos de historial
      if (history && history.length >= 2) {
        const prevRaw = history[history.length - 2];
        const currRaw = history[history.length - 1];

        const prev = Array.isArray(prevRaw)
          ? prevRaw
          : [prevRaw, Number.isFinite(det.center?.[1]) ? det.center[1] : this.lineY];
        const curr = Array.isArray(currRaw)
          ? currRaw
          : [currRaw, Number.isFinite(det.center?.[1]) ? det.center[1] : this.lineY];

        const prevDist = this._signedDistanceToLine(prev);
        const currDist = this._signedDistanceToLine(curr);
        const prevProj = this._lineProjection(prev);
        const currProj = this._lineProjection(curr);

        const halfLen = this.lineLengthPx / 2;
        const midProj = (prevProj + currProj) / 2;
        const insideSegment = Math.abs(prevProj) <= halfLen
          || Math.abs(currProj) <= halfLen
          || Math.abs(midProj) <= halfLen;

        const eps = 1.5;

        if (prevDist < -eps && currDist >= eps && state.canCross && insideSegment) {
          // Lado negativo → lado positivo = Entrada
          this.counts[categoria].in++;
          state.canCross = false;
          if (this.onCross) this.onCross(categoria, 'in', trackId);

        } else if (prevDist > eps && currDist <= -eps && state.canCross && insideSegment) {
          // Lado positivo → lado negativo = Salida
          this.counts[categoria].out++;
          state.canCross = false;
          if (this.onCross) this.onCross(categoria, 'out', trackId);

        } else if (Math.abs(currDist) > this.margin) {
          // Se alejó de la línea → puede volver a cruzar
          state.canCross = true;
        }
      }
    }

    // Limpiar tracks que ya no están activos
    for (const [id] of this._trackState) {
      if (!activeIds.has(id)) this._trackState.delete(id);
    }
  }

  /* ── Consultas ──────────────────────────────────────────────────────── */

  /** Devuelve copia de conteos con campo 'total' */
  getCounts() {
    const result = {};
    for (const [k, v] of Object.entries(this.counts)) {
      result[k] = { in: v.in, out: v.out, total: v.in + v.out };
    }
    return result;
  }

  /** Reinicia contadores y estados de cruce */
  reset() {
    CATEGORY_KEYS.forEach(k => { this.counts[k] = { in: 0, out: 0 }; });
    this._trackState.clear();
  }
}

File: counter.py
# -*- coding: utf-8 -*-
"""
counter.py — Lógica de conteo bidireccional.
Recibe detecciones con track_id (ya asignado por Ultralytics ByteTrack)
y detecta cruces de línea vertical.
"""

import threading
import logging
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Callable, Tuple

from config import CATEGORY_KEYS, LINE_POSITION_DEFAULT, CROSSING_MARGIN

logger = logging.getLogger(__name__)


@dataclass
class TrackInfo:
    track_id: int
    categoria: str
    historial_x: List[float] = field(default_factory=list)
    cruzado: bool = False
    puede_cruzar: bool = True
    ultimo_lado: str = "ninguno"  # 'izquierda' | 'derecha' | 'ninguno'


class BidirectionalCounter:
    """
    Contador bidireccional que trabaja con los track_id de Ultralytics.
    Detecta el cruce de la línea vertical por posición del centro del bbox.
    """

    def __init__(
        self,
        frame_width: int = 640,
        frame_height: int = 480,
        line_position: float = LINE_POSITION_DEFAULT,
        margin: int = CROSSING_MARGIN,
        on_cross: Optional[Callable] = None,
    ):
        self.frame_width  = frame_width
        self.frame_height = frame_height
        self.margin       = margin
        self.on_cross     = on_cross   # callback(categoria, dir_str, track_id)
        self._lock        = threading.Lock()

        self.line_x: int  = int(frame_width * line_position)
        self.line_pos: float = line_position

        self._tracks: Dict[int, TrackInfo] = {}

        # { "persons": {"in": 0, "out": 0}, ... }
        self.counts: Dict[str, Dict[str, int]] = {
            k: {"in": 0, "out": 0} for k in CATEGORY_KEYS
        }

    # ── Configuración dinámica ─────────────────────────────────────────────────

    def set_line_position(self, position: float) -> None:
        """Actualiza posición relativa de la línea (0.0 – 1.0)."""
        position = max(0.02, min(0.98, position))
        with self._lock:
            self.line_pos = position
            self.line_x   = int(self.frame_width * position)

    def update_frame_size(self, width: int, height: int) -> None:
        with self._lock:
            self.frame_width  = width
            self.frame_height = height
            self.line_x       = int(width * self.line_pos)

    # ── Procesamiento de detecciones ───────────────────────────────────────────

    def process(self, detections: List[Dict]) -> None:
        """
        Recibe lista de detecciones con campos:
            track_id, categoria, centro (cx, cy)
        Actualiza contadores si se detecta un cruce.
        """
        with self._lock:
            active_ids = set()
            for det in detections:
                tid = det.get("track_id")
                if tid is None:
                    continue
                categoria = det.get("categoria")
                if categoria not in self.counts:
                    continue
                cx, _cy = det["centro"]
                active_ids.add(tid)
                self._update_track(tid, categoria, cx)

            self._cleanup(active_ids)

    def _update_track(self, tid: int, categoria: str, cx: float) -> None:
        if tid not in self._tracks:
            self._tracks[tid] = TrackInfo(track_id=tid, categoria=categoria)

        track = self._tracks[tid]
        track.historial_x.append(cx)
        if len(track.historial_x) > 60:
            track.historial_x = track.historial_x[-60:]

        self._check_crossing(track)

    def _check_crossing(self, track: TrackInfo) -> None:
        if len(track.historial_x) < 2:
            return

        prev_x = track.historial_x[-2]
        curr_x = track.historial_x[-1]
        line   = self.line_x

        if prev_x < line and curr_x >= line and track.puede_cruzar:
            # left → right  (entrada = "in")
            self.counts[track.categoria]["in"] += 1
            track.puede_cruzar = False
            track.ultimo_lado  = "derecha"
            if self.on_cross:
                self.on_cross(track.categoria, "in", track.track_id)

        elif prev_x > line and curr_x <= line and track.puede_cruzar:
            # right → left  (salida = "out")
            self.counts[track.categoria]["out"] += 1
            track.puede_cruzar = False
            track.ultimo_lado  = "izquierda"
            if self.on_cross:
                self.on_cross(track.categoria, "out", track.track_id)

        else:
            # Resetear cuando el objeto se aleja de la línea
            if abs(curr_x - line) > self.margin:
                track.puede_cruzar = True

    def _cleanup(self, active_ids: set) -> None:
        """Elimina tracks que ya no aparecen en el frame."""
        stale = [tid for tid in self._tracks if tid not in active_ids]
        for tid in stale:
            del self._tracks[tid]

    # ── Consultas ──────────────────────────────────────────────────────────────

    def get_counts(self) -> Dict[str, Dict[str, int]]:
        """Devuelve copia de los contadores actuales con campo 'total' añadido."""
        with self._lock:
            result = {}
            for k, v in self.counts.items():
                result[k] = {"in": v["in"], "out": v["out"], "total": v["in"] + v["out"]}
            return result

    def reset(self) -> None:
        with self._lock:
            self._tracks.clear()
            for k in CATEGORY_KEYS:
                self.counts[k] = {"in": 0, "out": 0}
        logger.info("Contadores reiniciados")

    def get_line_position(self) -> Tuple[int, float]:
        """Devuelve (line_x_pixels, line_pos_relative)."""
        return self.line_x, self.line_pos
